src/HealthChecks/HealthChecks/src/HealthCheckPublisherHostedService.cs

```diff
@@ -17,36 +19,58 @@ namespace Microsoft.Extensions.Diagnostics.HealthChecks;
 {
     private readonly HealthCheckService _healthCheckService;
-    private readonly IOptions<HealthCheckPublisherOptions> _options;
+    private readonly IOptions<HealthCheckServiceOptions> _healthCheckServiceOptions;
+    private readonly IOptions<HealthCheckPublisherOptions> _healthCheckPublisherOptions;
     private readonly ILogger _logger;
     private readonly IHealthCheckPublisher[] _publishers;
+    private readonly (TimeSpan Delay, TimeSpan Period) _defaultTimerOptions;
+    private readonly Dictionary<(TimeSpan Delay, TimeSpan Period), List<HealthCheckRegistration>> _healthChecksByOptions;
+    private Dictionary<(TimeSpan Delay, TimeSpan Period), Timer>? _timersByOptions;
 
     private readonly CancellationTokenSource _stopping;
-    private Timer? _timer;
     private CancellationTokenSource? _runTokenSource;
 
     public HealthCheckPublisherHostedService(
         HealthCheckService healthCheckService,
-        IOptions<HealthCheckPublisherOptions> options,
+        IOptions<HealthCheckServiceOptions> healthCheckServiceOptions,
+        IOptions<HealthCheckPublisherOptions> healthCheckPublisherOptions,
         ILogger<HealthCheckPublisherHostedService> logger,
         IEnumerable<IHealthCheckPublisher> publishers)
     {
         ArgumentNullThrowHelper.ThrowIfNull(healthCheckService);
-        ArgumentNullThrowHelper.ThrowIfNull(options);
+        ArgumentNullThrowHelper.ThrowIfNull(healthCheckServiceOptions);
+        ArgumentNullThrowHelper.ThrowIfNull(healthCheckPublisherOptions);
         ArgumentNullThrowHelper.ThrowIfNull(logger);
         ArgumentNullThrowHelper.ThrowIfNull(publishers);
 
         _healthCheckService = healthCheckService;
-        _options = options;
+        _healthCheckServiceOptions = healthCheckServiceOptions;
+        _healthCheckPublisherOptions = healthCheckPublisherOptions;
         _logger = logger;
         _publishers = publishers.ToArray();
 
         _stopping = new CancellationTokenSource();
+
+        // We're specifically going out of our way to do this at startup time. We want to make sure you
+        // get any kind of health-check related error as early as possible. Waiting until someone
+        // actually tries to **run** health checks would be real baaaaad.
+        ValidateRegistrations(_healthCheckServiceOptions.Value.Registrations);
+
+        _defaultTimerOptions = (_healthCheckPublisherOptions.Value.Delay, _healthCheckPublisherOptions.Value.Period);
+
+        // Group healthcheck registrations by Delay, Period and Timeout, to build a Dictionary<(TimeSpan, TimeSpan, TimeSpan), List<HealthCheckRegistration>>
+        // For HCs with no Delay, Period or Timeout, we default to the publisher values
+        _healthChecksByOptions = _healthCheckServiceOptions.Value.Registrations.GroupBy(GetTimerOptionsOrDefault).ToDictionary(g => g.Key, g => g.ToList());
+    }
+
+    private (TimeSpan Delay, TimeSpan Period) GetTimerOptionsOrDefault(HealthCheckRegistration registration)
```

Comment:
I thought this might sometimes return `default` based on the name. Let's shorten it to `GetTimerOptions`.